services/purchase_request_copier.py

```diff
@@ -0,0 +1,163 @@
+        # ignoring auto-generated fields
+        elif field_data.field_def.type in [
+            "equation",
+            "concatenation",
+            "auto_increment",
+        ]:
+            continue
+        # All other record types
+        else:
+            output_record[field_name] = knack_record.get(field_name)
+
+    return output_record
+
+
+def assign_requester(data, requester_field_id, copied_by_field_id):
+    """
+    Makes the requester the person who created the copied record.
+    """
+    copied_by = data.pop(copied_by_field_id)
+    data[requester_field_id] = copied_by
+    return data
+
+
+def main(args):
+    # Process arguments
+    app_name = args.app_name
+    container = args.container
+    logger.info(args)
+    config = CONFIG[app_name][container]
+
+    # Use a "free" API call to check for if there's purchase requests to be copied
+    app = knackpy.App(app_id=APP_ID, api_key=None)
+    records = app.get(container, filters=None)
+
+    if len(records) == 0:
+        logger.info("No purchase requests to copy, did nothing.")
+        return 0
+
+    logger.info(f"Copying {len(records)} Purchase Requests")
+    # Creating a copy of the purchase request and assigning it to person requesting the copy
+    for purchase_request in records:
+        unique_id = purchase_request.get(config["unique_id_field_id"])
+        logger.info(f"Copying Purchase Request with ID: {unique_id}")
+        data = handle_record_types(purchase_request)
+        data = assign_requester(
+            data, config["requester_field_id"], config["copied_by_field_id"]
+        )
+
+        # set copy PR to "No"
+        data[config["copy_field_id"]] = False
+
+        # Create new record that was copied
+        copied_record = knackpy.api.record(
+            app_id=APP_ID,
+            api_key=API_KEY,
+            obj=config["object"],
+            method="create",
+            data=data,
+        )
+        logger.info(
+            f"New Purchase Request record generated with ID: {copied_record.get(config['unique_id_field_id'])}"
+        )
+
+        # Update the original copy to remove it from our queue of requested copies
+        old_record = {"id": purchase_request.get("id"), config["copy_field_id"]: False}
+        original_record = knackpy.api.record(
+            app_id=APP_ID,
+            api_key=API_KEY,
+            obj=config["object"],
+            method="update",
+            data=old_record,
+        )
+
+        # Get the Auto increment field ID, so we can search the PR items table
+        item_filter = {
+            "match": "and",
+            "rules": [
+                {
+                    "field": config["pr_items"]["pr_field_id"],
+                    "operator": "is",
+                    "value": unique_id,
+                }
+            ],
+        }
+        app = knackpy.App(app_id=APP_ID, api_key=API_KEY)
+        item_records = app.get(config["pr_items"]["object"], filters=item_filter)
```

Comment:
it would be nice to re-use the `App` instance here since you already have the metadata downloaded. 

I haven't tested this feature in a long time, but you should be able to pass the `refresh` flag to cause knackpy to fetch records anew. This feature is buried in the  [API docs](https://cityofaustin.github.io/knackpy/docs/api-reference/app/#get):

```python
 item_records = app.get(config["pr_items"]["object"], filters=item_filter, refresh=True)
```

another way to do that would be to construct the new instance with the metadata from the previous instance:

```python
app = knackpy.App(app_id=app_id,  metadata=app.metadata)
```

Reply:
I kinda wrapped your two comments together. Using the knackpy.api.get at the start and then authenticating with the API key above the first loop so we do not need to fetch the knack metadata more than once.